pilot/pkg/model/gateway_test.go

```diff
@@ -0,0 +1,268 @@
+					Tls:   tlsOne,
+					Hosts: []string{"example.com"},
+				}, {
+					Port:  port443,
+					Tls:   tlsTwo,
+					Hosts: []string{"example.com"},
+				}}},
+			nil,
+		},
+		{"same ports different domains",
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"bar.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"bar.com", "foo.com"},
+				},
+			}},
+			nil,
+		},
+		{"different domains, different ports",
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port443,
+					Tls:   tlsTwo,
+					Hosts: []string{"bar.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				}, {
+					Port:  port443,
+					Tls:   tlsTwo,
+					Hosts: []string{"bar.com"},
+				}}},
+			nil,
+		},
+		{"conflicting port names",
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80,
+					Tls:   tlsOne,
+					Hosts: []string{"foo.com"},
+				},
+			}},
+			&routing.Gateway{Servers: []*routing.Server{
+				{
+					Port:  port80DifferentName,
+					Tls:   tlsOne,
+					Hosts: []string{"bar.com"},
+				},
+			}},
+			nil,
+			errors.New(`unable to merge gateways: conflicting ports`),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// we want to save the original state of tt.a for printing if we fail the test, so we'll merge into a new gateway struct.
+			actual := &routing.Gateway{}
+			MergeGateways(actual, tt.a)
+			err := MergeGateways(actual, tt.b)
+			if err != tt.expectedErrPrefix && !strings.HasPrefix(err.Error(), tt.expectedErrPrefix.Error()) {
+				t.Fatalf("%s: got err %v, wanted %v", tt.name, err, tt.expectedErrPrefix)
+			}
+			if err == nil {
+				if !reflect.DeepEqual(actual, tt.expectedOut) {
+					t.Fatalf("%s: got %v, wanted %v", tt.name, actual, tt.expectedOut)
+				}
+			}
+		})
+	}
+
+}
+
+func TestPortsEqualAndDistinct(t *testing.T) {
+	tests := []struct {
+		name             string
+		b                *routing.Port
+		a                *routing.Port
+		expectedEqual    bool
+		expectedDistinct bool
+	}{
+		{"empty", &routing.Port{}, &routing.Port{}, true, false},
+		{"happy",
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			true, false},
+		{"same numbers but different names (case sensitive)",
+			&routing.Port{Number: 1, Name: "Bill", Protocol: "HTTP"},
+			&routing.Port{Number: 1, Name: "bill", Protocol: "HTTP"},
+			false, false},
```

Comment:
Shouldn’t this be false, true ?

Reply:
distinct means different ports and different names, in other words no ambiguity in listeners or in which route rules apply.  this notion will narrow in scope once we add support for SNI and TLS snooping, and merge h1/h2/grpc like you suggest.